internal/controller/apiextensions/composite/composition_functions.go

```diff
@@ -498,6 +505,29 @@ func (c *FunctionComposer) Compose(ctx context.Context, xr *composite.Unstructur
 	return CompositionResult{ConnectionDetails: d.GetComposite().GetConnectionDetails(), Composed: resources, Events: events}, nil
 }
 
+// ComposedFieldOwnerName generates a unique field owner name
+// for a given Crossplane composite resource (XR). This uniqueness is crucial to
+// prevent multiple XRs, which compose the same resource, from continuously
+// alternating as controllers.
+//
+// The function generates a deterministic hash based on the XR's name and
+// GroupKind (GK), ensuring consistency even during system restores. The hash
+// does not include the XR's UID (as it's not deterministic), namespace (XRs
+// don't have one), or version (to allow version changes without needing to
+// update the field owner name).
+//
+// We decided to include the GK in the hash to prevent transferring ownership of
+// composed resources across XRs with whole new GK, as that should not be
+// supported without manual intervention.
+//
+// Given that field owner names are limited to 128 characters, the function
+// truncates the hash to 32 characters. A longer hash was deemed unnecessary.
+func ComposedFieldOwnerName(xr *composite.Unstructured) string {
+	h := sha256.New()
+	_, _ = h.Write([]byte(xr.GetName() + xr.GroupVersionKind().GroupKind().String()))
+	return FieldOwnerComposedPrefix + "/" + hex.EncodeToString(h.Sum(nil))[:32]
```

Comment:
```suggestion
	return fmt.Sprintf("%s/%x", FieldOwnerComposedPrefix, h.Sum(nil))
```

FWIW if you don't truncate the hash at all the field owner will always fit. It will always be 101 characters, vs 69 when truncated. See https://go.dev/play/p/kTU-b-C67MK for example.

I don't feel strongly either way. The shorter hash is unlikely to collide assuming tens of thousands of XRs in a control plane, and looks a little better readability wise. On the other hand the longer hash will always fit, and gives us even more collision resistance. 🤷‍♂️

Reply:
went with this 👍